sentry-apollo-3/src/main/java/io/sentry/apollo3/SentryApollo3HttpInterceptor.kt

```diff
@@ -0,0 +1,162 @@
+package io.sentry.apollo3
+
+import com.apollographql.apollo3.api.http.HttpRequest
+import com.apollographql.apollo3.api.http.HttpResponse
+import com.apollographql.apollo3.exception.ApolloNetworkException
+import com.apollographql.apollo3.network.http.HttpInterceptor
+import com.apollographql.apollo3.network.http.HttpInterceptorChain
+import io.sentry.Breadcrumb
+import io.sentry.Hint
+import io.sentry.HubAdapter
+import io.sentry.IHub
+import io.sentry.ISpan
+import io.sentry.SentryLevel
+import io.sentry.SpanStatus
+import io.sentry.TracingOrigins
+import io.sentry.TypeCheckHint
+import okio.Buffer
+import java.io.InputStreamReader
+
+class SentryApollo3HttpInterceptor(private val hub: IHub = HubAdapter.getInstance(), private val beforeSpan: BeforeSpanCallback? = null) :
+    HttpInterceptor {
+
+    override suspend fun intercept(
+        request: HttpRequest,
+        chain: HttpInterceptorChain
+    ): HttpResponse {
+        val activeSpan = hub.span
+        return if (activeSpan == null) {
+            chain.proceed(request)
+        } else {
+            val span = startChild(request, activeSpan)
+
+            val requestBuilder = request.newBuilder()
+
+            if (TracingOrigins.contain(hub.options.tracingOrigins, request.url)) {
+                val sentryTraceHeader = span.toSentryTrace()
+                val baggageHeader = span.toBaggageHeader()
+                requestBuilder.addHeader(sentryTraceHeader.name, sentryTraceHeader.value)
+
+                baggageHeader?.let {
+                    requestBuilder.addHeader(it.name, it.value)
+                }
+            }
+
+            val modifiedRequest = requestBuilder.build()
+
+            return try {
+                val httpResponse = chain.proceed(modifiedRequest)
+                span.status = SpanStatus.fromHttpStatusCode(httpResponse.statusCode, SpanStatus.UNKNOWN)
+                finish(span, modifiedRequest, httpResponse)
+
+                httpResponse
+            } catch (e: Exception) {
+                when (e) {
+                    is ApolloNetworkException -> span.status = SpanStatus.INTERNAL_ERROR
+                    else -> SpanStatus.UNKNOWN
+                }
+                finish(span, modifiedRequest)
```

Comment:
Wouldn't be better to use a `finally` block that calls `finish`? otherwise, we duplicate code.

Reply:
Is it the two calls to `finish` that are bugging you?

`finish(span, modifiedRequest, httpResponse)` // Success i.e. there is a response from the server
`finish(span, modifiedRequest)` // Error i.e. no response from the server, or a local exception like no connectivity

The alternative would look like this:
```
var response: HttpResponse? = null

response = try {
    chain.proceed(modifiedRequest)
} catch (e: Throwable) {
    when (e) {
        is ApolloNetworkException -> span.status = SpanStatus.INTERNAL_ERROR
        else -> SpanStatus.UNKNOWN
    }
    throw e
} finally {
    span.status = SpanStatus.fromHttpStatusCode(response?.statusCode, SpanStatus.UNKNOWN)
    finish(span, modifiedRequest, response)
}

return response!!
```

Which, tbh, I don't like much. Especially since we need to force unwrap `response` at the end.

wdyt?